src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+        /// </remarks>
+        public static bool TryGetRuneAt(string input, int index, out Rune value)
+        {
+            if (input == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.input);
+            }
+
+            if ((uint)index >= (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+            }
+
+            // Optimistically assume input is within BMP.
+
+            uint returnValue = input[index];
+            if (UnicodeUtility.IsSurrogateCodePoint(returnValue))
+            {
+                if (!UnicodeUtility.IsHighSurrogateCodePoint(returnValue))
+                {
+                    goto Fail;
+                }
+
+                // Treat 'returnValue' as the high surrogate.
+                //
+                // If this becomes a hot code path, we can skip the below bounds check by reading
+                // off the end of the string using unsafe code. Since strings are null-terminated,
+                // we're guaranteed not to read a valid low surrogate, so we'll fail correctly if
+                // the string terminates unexpectedly.
+
+                index++;
+                if ((uint)index >= (uint)input.Length)
+                {
+                    goto Fail; // not an argument exception - just a "bad data" failure
+                }
+
+                uint potentialLowSurrogate = input[index];
+                if (!UnicodeUtility.IsLowSurrogateCodePoint(potentialLowSurrogate))
+                {
+                    goto Fail;
+                }
+
+                returnValue = UnicodeUtility.GetScalarFromUtf16SurrogatePair(returnValue, potentialLowSurrogate);
+            }
+
+            value = UnsafeCreate(returnValue);
+            return true;
+
+        Fail:
+            value = default;
+            return false;
+        }
+
+        // Allows constructing a Unicode scalar value from an arbitrary 32-bit integer without
+        // validation. It is the caller's responsibility to have performed manual validation
+        // before calling this method. If a Rune instance is forcibly constructed
+        // from invalid input, the APIs on this type have undefined behavior, potentially including
+        // introducing a security hole in the consuming application.
+        //
+        // An example of a security hole resulting from an invalid Rune value, which could result
+        // in a stack overflow.
+        //
+        // public int GetMarvin32HashCode(Rune r) {
+        //   Span<char> buffer = stackalloc char[r.Utf16SequenceLength];
+        //   s.TryEncode(buffer, ...);
+        //   return Marvin32.ComputeHash(buffer.AsBytes());
+        // }
+
+        /// <summary>
+        /// Creates a <see cref="Rune"/> without performing validation on the input.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Is this needed?  I'm surprised this isn't otherwise inlined.

Reply:
TBH I didn't measure that specifically. It was just a force of habit that I wrote it. :/